File: julia/src/interpolation/natural_cubic_spline.jl
module NaturalCubicSpline

export natural_cubic_spline

"""
        natural_spline(x_nodes, y_nodes, x)

Natural cubic spline interpolation.
Returns the interpolated value S(x).
"""

function natural_spline(x_nodes::Vector{<:Real}, y_nodes::Vector{<:Real}, x::Real)
    n = length(x_nodes)
    n == length(y_nodes) || error("x_nodes and y_nodes must have the same length.")
    n > 3 || error("At least 3 nodes are expected for the cubic spline.")

    a = copy(y_nodes)
    h = [x_nodes[i+1] - x_nodes[i] for i in 1:n-1]

    alpha = zeros(Float64, n)
    for i in 2:n-1
        alpha[i] = (3/h[i]) * (a[i+1] - a[i]) - (3/h[i-1]) * (a[i] - a[i-1])
    end

    l = zeros(Float64, n)
    mu = zeros(Float64, n)
    z = zeros(Float64, n)

    c = zeros(Float64, n)
    b = zeros(Float64, n)
    d = zeros(Float64, n)

    l[1] = 1.0
    
    for i in 2:n-1
        l[i] = 2*(x_nodes[i+1] - x_nodes[i-1]) - h[i-1]*mu[i-1]
        mu[i] = h[i] / l[i]
        z[i] = (alpha[i] - h[i-1]*z[i-1]) / l[i]
    end

    l[n] = 1.0

    for i in (n-1):-1:1
        c[i] = z[i] - mu[i] * c[i+1]
        b[i] = (a[i+1] - a[i]) / h[i] - h[i] * (c[i+1] + 2*c[i]) / 3
        d[i] = (c[i+1] - c[i]) / (3*h[i])
    end

    i = n - 1
    while i > 1 && x < x_nodes[i]
        i -= 1
    end

    dx = x - x_nodes[i]
    return a[i] + b[i]*dx + c[i]*dx^2 + d[i]*dx^3
end

if abspath(PROGRAM_FILE) == @__FILE__
    xs = [0.0, 1.0, 2.0, 3.0]
    ys = [0.0, 1.0, 0.0, 1.0]
    print(natural_spline(xs, ys, 1.5))
end

end                

File: fortran/src/interpolation/natural_cubic_spline.f90
module natural_cubic_spline mod
  implicit none
contains

  function natural_cubic_spline(x_nodes, y_nodes, x, n) result(Sx)
    implicit none
    integer, intent(in) :: n
    real(8), intent(in) :: x_nodes(n), y_nodes(n)
    real(8), intent(in) :: x
    real(8) :: Sx

    real(8) :: a(n), b(n), c(n), d(n)
    real(8) :: h(n-1), alpha(n), l(n), mu(n), z(n)
    integer :: i
    real(8) :: dx

    a = y_nodes

    do i = 1, n-1
       h(i) = x_nodes(i+1) - x_nodes(i)
    end do

    alpha(1) = 0d0
    alpha(n) = 0d0

    do i = 2, n-1
       alpha(i) = 3d0 * ( (a(i+1)-a(i))/h(i) - (a(i)-a(i-1))/h(i-1))
    end do

    l(1) = 1d0
    mu(1) = 0d0
    z(1) = 0d0
    c(1) = 0d0

    do i = 2, n-1
       l(i) = 2d0*(x_nodes(i+1) - x_nodes(i-1)) - h(i-1)*mu(i-1)
       mu(i) = h(i) / l(i)
    end do

    l(n) = 1d0
    z(n) = 0d0
    c(n) = 0d0

    do i = n-1, 1, -1
       c(i) = z(i) - mu(i)*c(i+1)
    end do

    do i = 1, n-1
       b(i) = (a(i+1)-a(i))/h(i) - h(i)*(2d0*c(i) + c(i+1))/3d0
       d(i) = (c(i+1) - c(i)) / (3d0*h(i))
    end do

    i = n-1
    do while (i > 1 .and. x < x_nodes(i))
       i = i - 1
    end do

    dx = x - x_nodes(i)

    Sx = a(i) + b(i)*dx + c(i)*dx*dx + d(i)*dx*dx*dx

  end function natural_cubic_spline

end module natural_cubic_spline
